### PR TITLE
Add delete function and add tests

### PR DESCRIPTION
**Pull request by amityadav0**

### CODE REVIEW

File: packages/algorand-js/src/runtime/runtime.ts [hunk 1]
```diff
@@ -318,6 +318,16 @@ export class Runtime {
     }
   }
 
+  // Delete application from account's state and global state
```
Comment (ratik21) [inline packages/algorand-js/src/runtime/runtime.ts:321]:
```suggestion
  // Delete application from account's local state and global state
```

File: packages/algorand-js/src/runtime/runtime.ts [hunk 2]
```diff
@@ -347,12 +357,18 @@ export class Runtime {
...
+        }
       }
     } else {
       // for a single (stand alone) transaction
       this.store.accounts.forEach((account, addr) => {
         this.updateBalance(txnParams, account);
       });
+      if (txnParams.type === TransactionType.DeleteSSC) {
+        this.deleteApp(txnParams.fromAccount.addr, txnParams.appId);
+      }
```
Comment (ratik21) [inline packages/algorand-js/src/runtime/runtime.ts:371]: We have repetitive code in this function (which will probably increase when we add more txn types). I would suggest to:
+ either move the common code to a function and reuse the function. OR
+ treat the single txn as a group of 1 txn (`const tx = [txnParams];`), and just operate the updates on an array once.

File: packages/algorand-js/test/integration/deleteApp.ts [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+  it("should delete application", async function () {
+    const appId = await runtime.addApp(flags, {}, program);
+    deleteParams.appId = appId;
+    await runtime.executeTx(deleteParams, program, []);
+
+    // verify app is deleted
+    const res = runtime.ctx.state.globalApps.has(appId);
+    assert.equal(res, false);
+  });
+});
```
Comment (ratik21) [inline packages/algorand-js/test/integration/deleteApp.ts:54]: Please add one more test: 
+ create an app
+ execute transaction which should be rejected by logic (maybe we need to modify teal code)
+ assert that the app is not deleted as txn is rejected.

File: packages/algorand-js/src/runtime/runtime.ts [hunk 4]
```diff
@@ -318,6 +318,16 @@ export class Runtime {
     }
   }
 
+  // Delete application from account's state and global state
+  deleteApp (accountAddr: string, appId: number): void {
+    if (!this.store.globalApps.has(appId)) {
+      throw new TealError(ERRORS.TEAL.APP_NOT_FOUND);
+    }
+    this.store.globalApps.delete(appId);
+    const account = this.assertAccountDefined(this.store.accounts.get(accountAddr));
```
Comment (ratik21) [inline packages/algorand-js/src/runtime/runtime.ts:327]: What happens if we delete the `app` and account is not defined?

File: packages/algorand-js/src/runtime/runtime.ts (hunk 1 above)
Comment (amityadav0) [reply packages/algorand-js/src/runtime/runtime.ts:321]: we are deleting it from global too

File: packages/algorand-js/src/runtime/runtime.ts (hunk 2 above)
Comment (amityadav0) [reply packages/algorand-js/src/runtime/runtime.ts:371]: second one seems good

File: packages/algorand-js/src/runtime/runtime.ts (hunk 4 above)
Comment (amityadav0) [reply packages/algorand-js/src/runtime/runtime.ts:327]: Nice catch

File: packages/algorand-js/test/integration/deleteApp.ts (hunk 3 above)
Comment (amityadav0) [reply packages/algorand-js/test/integration/deleteApp.ts:54]: thanks for suggestion

**Comment by ratik21** [comment]
@amityadav0 the current code is only deleting application from global state (creator's account). Can you please check one thing: Let's say we have 3 accounts A, B, C. A has created an app and A, B, C opt-in to that app. Now if we delete the app, what happens to the `localState` of A, B, C? Does the app get's deleted from there as well? If yes then we need to modify the code. Can you check?

File: packages/algorand-js/src/runtime/runtime.ts [hunk 5]
```diff
@@ -342,17 +356,19 @@ export class Runtime {
    * @param accounts : accounts passed by user
```
Comment (robert-zaremba) [inline packages/algorand-js/src/runtime/runtime.ts:356]: there is no `accounts` parameter in this function. Could you update the function description? Please be more specific which account balances are we updating - it seams that we update all existing accounts.

File: packages/algorand-js/test/integration/crowdfunding.ts [hunk 6]
```diff
@@ -64,7 +63,7 @@ describe("Crowdfunding basic tests", function () {
     const appId = await runtime.addApp({ ...validFlags, appArgs: appArgs }, {}, program);
     const getGlobal = (key: string):
       bigint | Uint8Array |undefined => runtime.getGlobalState(appId, stringToBytes(key));
-    const johnPk = decodeAddress(john.address).publicKey;
+    const johnPk = addressToBytes(john.address);
```
Comment (robert-zaremba) [inline packages/algorand-js/test/integration/crowdfunding.ts:66]: how about renaming `addressToBytes` -> `addressToPk`? BTW, I thought that address doesn't have a reverse function to Public Key.

**Review by robert-zaremba — APPROVED** [review]
pre-approving.

**Comment by amityadav0** [comment]
@ratik21 i think you are talking about closeout and clear, we will add them when we add full support for application calls, i.e after payment checks and asset checks

**Comment by ratik21** [comment]
> @ratik21 i think you are talking about closeout and clear, we will add them when we add full support for application calls, i.e after payment checks and asset checks

Okay, understood. So that means delete call only delete's the app from creator's account.

File: packages/algorand-js/src/runtime/runtime.ts (hunk 5 above)
Comment (amityadav0) [reply packages/algorand-js/src/runtime/runtime.ts:356]: ok

Comment (amityadav0) [reply packages/algorand-js/src/runtime/runtime.ts:356]: yes, we are traversing a list of accounts it is not required, updating. thanks

File: packages/algorand-js/test/integration/crowdfunding.ts (hunk 6 above)
Comment (amityadav0) [reply packages/algorand-js/test/integration/crowdfunding.ts:66]: for int and string, we are using toBytes, and according to decodeAddress, it returns pk as Uint8Array. so looking from user perspective i think `addressToBytes` is ok(i added doc string addressToBytes about Public key).
should we rename it to `addressToPk` ?

Comment (amityadav0) [reply packages/algorand-js/test/integration/crowdfunding.ts:66]: updated to `addressToPk`

**Pull request merged**